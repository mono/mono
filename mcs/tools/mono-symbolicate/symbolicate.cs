using System;
using System.IO;
using System.Globalization;
using System.Text.RegularExpressions;

namespace Symbolicate
{
	public class Program
	{
		static Regex regex = new Regex (@"\w*at (?<Method>.+) *(\[0x(?<IL>.+)\]|<0x.+ \+ 0x(?<NativeOffset>.+)>( (?<MethodIndex>\d+)|)) in <filename unknown>:0");

		public static int Main (String[] args)
		{
			if (args.Length < 2) {
				Console.Error.WriteLine ("Usage: symbolicate <assembly path> <input file> [lookup directories]");
				return 1;
			}

			var assemblyPath = args [0];
			var inputFile = args [1];

			var locProvider = new LocationProvider ();

			for (var i = 2; i < args.Length; i++)
				locProvider.AddDirectory (args [i]);

			locProvider.AddAssembly (assemblyPath);

			using (StreamReader r = new StreamReader (inputFile)) {
			    for (var line = r.ReadLine (); line != null; line = r.ReadLine ()) {
					line = SymbolicateLine (line, locProvider);
					Console.WriteLine (line);
			    }
			}

			return 0;
		}

		static string SymbolicateLine (string line, LocationProvider locProvider)
		{
			var match = regex.Match (line);
			if (!match.Success)
				return line;

			string typeFullName;
			var methodStr = match.Groups ["Method"].Value.Trim ();
			if (!TryParseMethodType (methodStr, out typeFullName))
				return line;

			var isOffsetIL = !string.IsNullOrEmpty (match.Groups ["IL"].Value);
			var offsetVarName = (isOffsetIL)? "IL" : "NativeOffset";
			var offset = int.Parse (match.Groups [offsetVarName].Value, NumberStyles.HexNumber, CultureInfo.InvariantCulture);

			uint methodIndex = 0xffffff;
			if (!string.IsNullOrEmpty (match.Groups ["MethodIndex"].Value))
				methodIndex = uint.Parse (match.Groups ["MethodIndex"].Value, CultureInfo.InvariantCulture);

			Location location;
			if (!locProvider.TryGetLocation (methodStr, typeFullName, offset, isOffsetIL, methodIndex, out location))
				return line;

			return line.Replace ("<filename unknown>:0", string.Format ("{0}:{1}", location.FileName, location.Line));
		}

		static bool TryParseMethodType (string str, out string typeFullName)
		{
			typeFullName = null;

			var methodNameEnd = str.IndexOf ("(");
			if (methodNameEnd == -1)
				return false;

			// Remove parameters
			str = str.Substring (0, methodNameEnd);

			// Remove generic parameters
			str = Regex.Replace (str, @"\[[^\[\]]*\]", "");

			var typeNameEnd = str.LastIndexOf (".");
			if (methodNameEnd == -1 || typeNameEnd == -1)
				return false;

			// Remove method name
			typeFullName = str.Substring (0, typeNameEnd);

			return true;
		}
	}
}