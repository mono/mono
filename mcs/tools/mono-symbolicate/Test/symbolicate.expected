System.Exception: Stacktrace with 1 frame
  at StackTraceDumper.Main () in StackTraceDumper.cs:9 
Stacktrace:
  at StackTraceDumper.Main () in StackTraceDumper.cs:9 

System.Exception: Stacktrace with 2 frames
  at StackTraceDumper.<Main>m__0 () in StackTraceDumper.cs:16 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 
Stacktrace:
  at StackTraceDumper.<Main>m__0 () in StackTraceDumper.cs:16 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 

System.Exception: Stacktrace with 3 frames
  at StackTraceDumper.ThrowException (System.String message, System.Int32 i) in StackTraceDumper.cs:78 
  at StackTraceDumper.<Main>m__1 () in StackTraceDumper.cs:18 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 
Stacktrace:
  at StackTraceDumper.ThrowException (System.String message, System.Int32 i) in StackTraceDumper.cs:78 
  at StackTraceDumper.<Main>m__1 () in StackTraceDumper.cs:18 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 

System.Exception: Stacktrace with 4 frames
  at StackTraceDumper.ThrowException (System.String message, System.Int32 i) in StackTraceDumper.cs:78 
  at StackTraceDumper.ThrowException (System.String message, System.Int32 i) in StackTraceDumper.cs:76 
  at StackTraceDumper.<Main>m__2 () in StackTraceDumper.cs:20 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 
Stacktrace:
  at StackTraceDumper.ThrowException (System.String message, System.Int32 i) in StackTraceDumper.cs:78 
  at StackTraceDumper.ThrowException (System.String message, System.Int32 i) in StackTraceDumper.cs:76 
  at StackTraceDumper.<Main>m__2 () in StackTraceDumper.cs:20 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 

System.Exception: Stack frame with method overload using ref parameter
  at StackTraceDumper.ThrowException (System.String& message) in StackTraceDumper.cs:70 
  at StackTraceDumper.<Main>m__3 () in StackTraceDumper.cs:24 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 
Stacktrace:
  at StackTraceDumper.ThrowException (System.String& message) in StackTraceDumper.cs:70 
  at StackTraceDumper.<Main>m__3 () in StackTraceDumper.cs:24 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 

System.Exception: Stack frame with method overload using out parameter
  at StackTraceDumper.ThrowException (System.String message, System.Int32& o) in StackTraceDumper.cs:83 
  at StackTraceDumper.<Main>m__4 () in StackTraceDumper.cs:29 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 
Stacktrace:
  at StackTraceDumper.ThrowException (System.String message, System.Int32& o) in StackTraceDumper.cs:83 
  at StackTraceDumper.<Main>m__4 () in StackTraceDumper.cs:29 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 

System.Exception: Stack frame with 1 generic parameter
  at StackTraceDumper.ThrowExceptionGeneric[T] (System.String message) in StackTraceDumper.cs:88 
  at StackTraceDumper.<Main>m__5 () in StackTraceDumper.cs:32 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 
Stacktrace:
  at StackTraceDumper.ThrowExceptionGeneric[T] (System.String message) in StackTraceDumper.cs:88 
  at StackTraceDumper.<Main>m__5 () in StackTraceDumper.cs:32 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 

System.Exception: Stack frame with 2 generic parameters
  at StackTraceDumper.ThrowExceptionGeneric[T1,T2] (System.String message) in StackTraceDumper.cs:108 
  at StackTraceDumper.<Main>m__6 () in StackTraceDumper.cs:34 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 
Stacktrace:
  at StackTraceDumper.ThrowExceptionGeneric[T1,T2] (System.String message) in StackTraceDumper.cs:108 
  at StackTraceDumper.<Main>m__6 () in StackTraceDumper.cs:34 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 

System.Exception: Stack frame with generic method overload
  at StackTraceDumper.ThrowExceptionGeneric[T] (T a1) in StackTraceDumper.cs:93 
  at StackTraceDumper.<Main>m__7 () in StackTraceDumper.cs:36 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 
Stacktrace:
  at StackTraceDumper.ThrowExceptionGeneric[T] (T a1) in StackTraceDumper.cs:93 
  at StackTraceDumper.<Main>m__7 () in StackTraceDumper.cs:36 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 

System.Exception: Stack trace with inner class
  at StackTraceDumper+InnerClass.ThrowException (System.String message) in StackTraceDumper.cs:114 
  at StackTraceDumper.<Main>m__8 () in StackTraceDumper.cs:38 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 
Stacktrace:
  at StackTraceDumper+InnerClass.ThrowException (System.String message) in StackTraceDumper.cs:114 
  at StackTraceDumper.<Main>m__8 () in StackTraceDumper.cs:38 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 

System.Exception: Stack trace with inner generic class
  at StackTraceDumper+InnerGenericClass`1[T].ThrowException (System.String message) in StackTraceDumper.cs:121 
  at StackTraceDumper.<Main>m__9 () in StackTraceDumper.cs:40 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 
Stacktrace:
  at StackTraceDumper+InnerGenericClass`1[T].ThrowException (System.String message) in StackTraceDumper.cs:121 
  at StackTraceDumper.<Main>m__9 () in StackTraceDumper.cs:40 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 
Generic to string:string

System.Exception: Stack trace with inner generic class and method generic parameter
  at StackTraceDumper+InnerGenericClass`1[T].ThrowException (System.String message, T arg) in StackTraceDumper.cs:127 
  at StackTraceDumper.<Main>m__A () in StackTraceDumper.cs:42 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 
Stacktrace:
  at StackTraceDumper+InnerGenericClass`1[T].ThrowException (System.String message, T arg) in StackTraceDumper.cs:127 
  at StackTraceDumper.<Main>m__A () in StackTraceDumper.cs:42 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 

System.Exception: Stack trace with inner generic class and generic overload
  at StackTraceDumper+InnerGenericClass`1[T].ThrowException[T1] (System.String message, T1 arg) in StackTraceDumper.cs:132 
  at StackTraceDumper.<Main>m__B () in StackTraceDumper.cs:44 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 
Stacktrace:
  at StackTraceDumper+InnerGenericClass`1[T].ThrowException[T1] (System.String message, T1 arg) in StackTraceDumper.cs:132 
  at StackTraceDumper.<Main>m__B () in StackTraceDumper.cs:44 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 

System.Exception: Stack trace with 2 inner generic class and generic overload
  at StackTraceDumper+InnerGenericClass`1+InnerInnerGenericClass`1[T,T2].ThrowException (T message) in StackTraceDumper.cs:138 
  at StackTraceDumper.<Main>m__C () in StackTraceDumper.cs:46 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 
Stacktrace:
  at StackTraceDumper+InnerGenericClass`1+InnerInnerGenericClass`1[T,T2].ThrowException (T message) in StackTraceDumper.cs:138 
  at StackTraceDumper.<Main>m__C () in StackTraceDumper.cs:46 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 

System.Exception: Stack trace with 2 inner generic class and generic overload
  at StackTraceDumper+InnerGenericClass`1+InnerInnerGenericClass`1[T,T2].ThrowException (T2 message) in StackTraceDumper.cs:143 
  at StackTraceDumper.<Main>m__D () in StackTraceDumper.cs:48 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 
Stacktrace:
  at StackTraceDumper+InnerGenericClass`1+InnerInnerGenericClass`1[T,T2].ThrowException (T2 message) in StackTraceDumper.cs:143 
  at StackTraceDumper.<Main>m__D () in StackTraceDumper.cs:48 
  at StackTraceDumper.Catch (System.Action action) in StackTraceDumper.cs:54 
